_drafts/herencia-y-composición.md

```diff
@@ -0,0 +1,324 @@
+---
+layout: post
+title: Herencia y Composición
+author: juliogarciag
+tags:
+    - oop
+    - ruby
+---
+Antes de embarcarnos en explicar qué es la composición, hay que explicar el por qué de este post. La idea es entender *qué* es la composición y por qué existe la frase (extraída del libro [Design Patterns](https://en.wikipedia.org/wiki/Design_Patterns)):
+
+> Favor 'object composition' over 'class inheritance'.
+
+Pensemos primero en un problema: Tenemos una función de un chat que recibe un mensaje (quien sabe cómo aunque puede ser por websockets o por long polling) y debe notificarlo inmediatamente a varios lugares:
+
+```ruby
+class Notifier
+  def initialize(raw_message)
+    self.raw_message = raw_message
+  end
+
+  def send_notifications
+    send_by_websockets(create_event_for_websockets)
+    notify_third_party_service(raw_message)
+  end
+
+  private
+
+  attr_accessor :raw_message
+
+  def send_by_websockets
+    # ...
+  end
+
+  def create_event_for_websockets
+    # ...
+  end
+
+  # ... more private methods
+end
+```
+
+Unas semanas después, tenemos que implementar una feature super genial que lee frases tipo: i want 1 cup of coffee que sean enviadas a ciertos usuarios especiales (bots) y generen órdenes de compra de varios bienes y servicios. Sabemos que tenemos que parsear esa data y sabemos que esto puede resultar complicado, así que creamos un mixin:
+
+```ruby
+module OrderSender
+  def send_orders_from_message
+    # ...
+  end
+end
+
+class Notifier
+  include OrderSender
+
+  # ...
+
+  def send_notifications
+    send_orders_from_message
+    send_by_websockets(create_event_for_websockets)
+    notify_third_party_service(raw_message)
+  end
+
+  # ...
+end
+```
+
+Luego realizamos la feature en el módulo que fue incluido. Con el tiempo nos encontramos con que deberíamos testear esto. Si de todos modos el mensaje en raw es una cadena encodificada, entonces deberíamos ser capaces de realizar tests unitarios que nos ayuden a no romper ninguna funcionalidad y a descubrir posibles errores. Así que empezamos:
+
+
+
+```ruby
+describe OrderSender do
+  describe "#send_orders_from_message" do
+  end
+end
+```
+
+Y aquí nos detenemos un minuto: ¿Cómo testeamos un módulo? Tenemos que crear un objeto para incluirlo y luego poder usarlo, algo así:
+
+```ruby
+describe OrderSender do
+  describe "#send_orders_from_message" do
+    it "recognizes messages and set them to somewhere else" do
+      object = Object.new
+      object.send(:extend, OrderSender)
+    end
+  end
+end
+```
+
+Y en la prueba nos damos cuenta que nos hemos olvidado del `raw_message`, así que tenemos que crear una clase completa que pueda tener ese mensaje o usamos un doble:
```

Comment:
De todos modos prefiero usar éste porque es más cercano al rubyista y no requiere conocer otro lenguaje: https://semaphoreci.com/community/tutorials/mocking-with-rspec-doubles-and-expectations